src/OrchardCore.Modules/OrchardCore.Media/Services/ViewMediaFolderAuthorizationHandler.cs

```diff
@@ -0,0 +1,229 @@
+                }
+            }
+            else
+            {
+                // Authorize by using the content item permission. The user must have access to the content item to allow its media
+                // as well.
+                var contentItemId = attachedMediaPathParts.Length > 1 ? attachedMediaPathParts[1] : null;
+                var contentItem = !string.IsNullOrEmpty(contentItemId) ? await _contentManager.GetAsync(contentItemId) : null;
+
+                // Disallow if content item is not found or allowed
+                if (contentItem is not null)
+                {
+                    await AuthorizeAsync(context, requirement, Contents.CommonPermissions.ViewContent, contentItem);
+                }
+            }
+        }
+
+        private async Task AuthorizeUsersFolderAsync(AuthorizationHandlerContext context, PermissionRequirement requirement, string path)
+        {
+            // We need to allow the _Users folder for own media access too. If someone uploads into this folder, we are screwed.
+            Permission permission;
+            if (path.IndexOf(PathSeparator) < 0)
+            {
+                permission = SecureMediaPermissions.ViewOwnMedia;
+            }
+            else
+            {
+                permission = SecureMediaPermissions.ViewOthersMedia;
+
+                var userFolderName = _userAssetFolderNameProvider.GetUserAssetFolderName(context.User);
+                if (!string.IsNullOrEmpty(userFolderName))
+                {
+                    var userOwnFolder = EnsureTrailingSlash(_fileStore.Combine(_usersFolder, userFolderName));
+
+                    if (IsAuthorizedFolder(userOwnFolder, path) || IsDescendantOfAuthorizedFolder(userOwnFolder, path))
+                    {
+                        permission = SecureMediaPermissions.ViewOwnMedia;
+                    }
+                }
+            }
+
+            await AuthorizeAsync(context, requirement, permission);
+        }
+
+        private async Task AuthorizeAsync(AuthorizationHandlerContext context, PermissionRequirement requirement, Permission permission, object resource = null)
+        {
+            var authorizationService = _serviceProvider.GetService<IAuthorizationService>();
+            if (await authorizationService.AuthorizeAsync(context.User, permission, resource))
+            {
+                // If anonymous access is also possible, we want to use default browser caching policies.
+                // Otherwise we set a marker which causes a different caching policy being used.
+                if ((context.User.Identity?.IsAuthenticated ?? false) && !await authorizationService.AuthorizeAsync(_anonymous, permission, resource))
+                {
+                    _httpContextAccessor.HttpContext.MarkAsSecureMediaRequested();
+                }
+
+                context.Succeed(requirement);
+            }
```

Comment:
I get `Cache-Control: public, must-revalidate, max-age=2592000` for files in secured folders too, without anonymous access to that folder (anonymous only has a folder-specific permission for another folder, and it does NOT have "View media content in all folders"). Most possibly related, that anonymous can't access files in a folder they are supposed to have access to:


Uploading 2024-04-16_22h03_37.mp4…

Reply:
Hmm, yes there is an oddity here in general in Orchards permission system. For a site owner, the new authorization handler is skipped because the `PermissionHandler` will always succeed any the requirement if the `SiteOwner` permission has been granted. Because the authorization handler will not revoke that grant, default caching is always being applied. It is the same for other permissions, e.g. the `ManageMediaFolder` permission.

I think it is not that problematic, as a site owner will always have access anyway to everything. 

> Most possibly related, that anonymous can't access files in a folder they are supposed to have access to.

That I could not reproduce. Can you elaborate?

Reply:
The video I wanted to attach apparently didn't fully upload, so here it is again:

https://github.com/OrchardCMS/OrchardCore/assets/1976647/40382fe2-556a-47de-ae95-58d3240be872

The authenticated user I used in this was the default admin, so having site owner permission, yes.

Reply:
Oh, I see. Folders with spaces are buggy, I'll fix it.

Reply:
I think it's fixed.